File: src/main/scala/com/vc/hard/L527.scala
package com.vc.hard

object L527 {
  def wordsAbbreviation(dict: List[String]): List[String] = {
    import scala.collection.mutable
    val arr = new Array[(String, Int)](dict.length)

    //Generate a list with prefix = 1
    dict.indices.foreach(i => {
      val abbr = makeAbbr(dict(i), 1)
      arr(i) = (abbr, 1)
    })

    //Increase abbr prefix for those words whose abbr have conflict
    dict.indices.foreach(i => {
      resolveConflict(i)
    })

    def resolveConflict(from: Int): Unit = {
      while(true) {
        val set = new mutable.HashSet[Int]()
        (from + 1 until dict.length).foreach(j => {
          if(arr(from) == arr(j)) set += j
        })
        if(set.isEmpty) return
        set += from
        set.foreach(i => {
          val abbr = makeAbbr(dict(i), arr(i)._2 + 1)
          arr(i) = (abbr, arr(i)._2 + 1)
        })
      }
    }

    def makeAbbr(word: String, prefix: Int): String = {
      if(prefix >= word.length - 2) return word
      val sb = new mutable.StringBuilder()
      sb.append(word.substring(0, prefix))
      sb.append(word.length - 1 - prefix)
      sb.append(word(word.length - 1))
      return sb.toString
    }
    arr.map(x => x._1).toList
  }
}

File: src/main/java/com/vc/hard/L527.java
package com.vc.hard;

import java.util.*;

class L527 {

    static class Entry {
        String str;
        int index;
        int count;

        Entry(String str, int index, int count) {
            this.str = str;
            this.index = index;
            this.count = count;
        }
    }

    public List<String> wordsAbbreviation(List<String> dict) {
        List<Entry> input = new ArrayList<>();
        for(int i = 0; i < dict.size(); i++) {
            String abbr = getAbbr(dict.get(i), 1);
            input.add(new Entry(abbr, i, 1));
        }

        for(int i = 0; i < input.size(); i++) {
            resolveConflict(dict, input, i);
        }

        List<String> res = new ArrayList<>();
        for(int i = 0; i < input.size(); i++) {
            res.add(input.get(i).str);
        }
        return res;
    }

    private void resolveConflict(List<String> dict, List<Entry> input, int from) {
        while(true) {
            HashSet<Integer> set = new HashSet<>();
            for(int i = from + 1; i < input.size(); i++) {
                if(input.get(from).str.equals(input.get(i).str)) set.add(i);
            }
            if(set.isEmpty()) return;
            set.add(from);
            for(Integer i: set) {
                String abbr = getAbbr(dict.get(i), input.get(i).count + 1);
                input.set(i, new Entry(abbr, i, input.get(i).count + 1));
            }
        }
    }

    private String getAbbr(String word, int count) {
        int n = word.length();
        if(n <= 3) return word;
        else if(count >= n) return word;
        StringBuilder str = new StringBuilder();
        str.append(word.substring(0, count));
        str.append(n - count - 1);
        str.append(word.charAt(n - 1));
        if(str.length() >= word.length()) return word;
        return str.toString();
    }
}
